### PR TITLE
[C++] Use RayConfig to read internal environment variables only once

### LINKED ISSUE #18756: [C++] Make accessing environment variable safe from concurrent modifications
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

In very rare cases, Ray can crash on call `std::getenv()`, likely due to concurrent modifications to the environment variables. Although Ray does not seem to modify environment variables itself, the user code it runs can. One possible protection against that is to store all environment variables on core worker startup, and look up from that map later.

### Reproduction script

n/a

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
This is to protect against concurrent modifications of the environment by other logic, by just reading the environment once on worker startup (calling `GetProcessJobID()` in `core_worker.cc`). Ray does not modify environment itself, so it is unnecessary to re-read environment variables after a worker process has started.

## Related issue number
closes #18756

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: cpp/src/ray/config_internal.cc [hunk 1]
```diff
@@ -114,11 +116,11 @@ void ConfigInternal::Init(RayConfig &config, int argc, char **argv) {
   }
   if (worker_type == WorkerType::DRIVER && run_mode == RunMode::CLUSTER) {
     if (redis_ip.empty()) {
-      auto ray_address_env = std::getenv("RAY_ADDRESS");
+      const auto ray_address_env = GetEnvironment("RAY_ADDRESS");
```
Comment (iycheng) [inline cpp/src/ray/config_internal.cc:119]: How about let's just put it into:
https://github.com/ray-project/ray/blob/master/src/ray/common/ray_config_def.h

and use instance() for this?

Comment (ericl) [reply cpp/src/ray/config_internal.cc:119]: +1

Comment (ericl) [reply cpp/src/ray/config_internal.cc:119]: I think this should work as a pattern for all the vars. We can recommend to never use raw env vars directly, it has to go through ray_config_def.h

File: src/ray/util/process.cc [hunk 2]
```diff
@@ -60,6 +59,29 @@ int execvpe(const char *program, char *const argv[], char *const envp[]) {
...
+    RAY_CHECK(*e && **e != '\0') << "environment variable name is absent";
+    const char *key_end = strchr(*e + 1 /* +1 is needed for Windows */, '=');
+    RAY_CHECK(key_end) << "environment variable value is absent: " << e;
+    env[std::string(*e, static_cast<size_t>(key_end - *e))] = key_end + 1;
+  }
+  return env;
+}
+}  // namespace
+
+std::optional<std::string> GetEnvironment(const std::string &name) {
```
Comment (iycheng) [inline src/ray/util/process.cc:76]: should it be `std::optional<const std::string&>`?

File: cpp/src/ray/config_internal.cc (hunk 1 above)
Comment (mwtian) [reply cpp/src/ray/config_internal.cc:119]: I feel ray_config_def.h contains different types of configuration options, which allow users to browse and see what they can change, via env, flags or config files. Do ADDRESS, JOB_ID and RAYLET_PID really belong there? They are not intended to be set by anyone other than Ray. Maybe we can come up with a standard on propagating settings in future.

File: src/ray/util/process.cc (hunk 2 above)
Comment (mwtian) [reply src/ray/util/process.cc:76]: `std::optional` does not support reference type in clang on my macbook. I could use `std::string_view` but subsequent logic accepts `const std::string &`. I think copying a small string is fine here.

File: cpp/src/ray/config_internal.cc (hunk 1 above)
Comment (ericl) [reply cpp/src/ray/config_internal.cc:119]: Well, there's the machinery vs the name (e.g., RAY_CONFIG). If you feel strongly about it, how about defining a new macro like RAY_PROCESS_FLAG, which can use the same implementation as RAY_CONFIG under the hood? Then we can define, e.g.:

RAY_PROCESS_FLAG(std::string, JOB_ID, "")

In a separate file for organizational purposes, but still use the same parsing and init logic.

Comment (iycheng) [reply cpp/src/ray/config_internal.cc:119]: If we go this way, I'd suggest `RAY_INTERNAL_FLAG`

Comment (mwtian) [reply cpp/src/ray/config_internal.cc]: Both `RAY_INTERNAL_FLAG` and `RAY_PROCESS_FLAG` sound ok to me, maybe `RAY_INTERNAL_FLAG` / `RAY_INTERNAL_CONFIG` is slightly better. Let me know which is the most suitable and I will update the naming later.

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Hmm windows build failure

File: src/ray/util/process.cc (hunk 2 above)
Comment (iycheng) [reply src/ray/util/process.cc]: Should it be a `const std::string*`?

File: src/ray/common/ray_config.h [hunk 3]
```diff
@@ -63,7 +63,16 @@ class RayConfig {
   inline type &name() { return name##_; }
 
 #include "ray/common/ray_config_def.h"
-/// -------------------------------------------------------------------------
+
+/// -----------Include ray_process_flag_def.h to define internal flags-------
```
Comment (iycheng) [inline src/ray/common/ray_config.h:67]: How about RAY_INTERNAL_FLAG? From the comment, it's also saying it's internal flags I think?

My feeling is that RAY_PROCESS_FLAGS doesn't give a clear explanation about that it's for internal use and don't use it externally.

Comment (iycheng) [reply src/ray/common/ray_config.h:67]: Sorry, I missed your previous comments. I think something INTERNAL there looks better.

Comment (mwtian) [reply src/ray/common/ray_config.h:67]: Sounds good. Will update to INTERNAL

**Comment by mwtian** [comment]
Windows build is fixed. Will merge a change to fix the GPU docker issue.

File: src/ray/util/process.cc (hunk 2 above)
Comment (mwtian) [reply src/ray/util/process.cc]: This has been removed.

**Review by fishbone — APPROVED** [review]
LG. Please fix the test.

**Comment by scv119** [comment]
Thanks @mwtian for fixing this! 
As a follow up, do you think it makes sense that we forbid the use of std::getenv by clang-tidy by default?

**Comment by scv119** [comment]
linter is failing dues to clang-format-12 :)
also looks https://github.com/ray-project/ray/blob/master/src/ray/util/logging.cc#L173 and https://sourcegraph.com/github.com/ray-project/ray/-/blob/src/ray/util/filesystem.cc are using getenv too. 
wondering if we should go ahead and address them...

**Comment by mwtian** [comment]
> linter is failing dues to clang-format-12 :)
> also looks https://github.com/ray-project/ray/blob/master/src/ray/util/logging.cc#L173 and https://sourcegraph.com/github.com/ray-project/ray/-/blob/src/ray/util/filesystem.cc are using getenv too.
> wondering if we should go ahead and address them...

Currently all crashes are only observed in `CoreWorker::CheckForRayletFailure()`. For callsites of `getenv` / `std::getenv` under `ray/util`, they are only called on process start up, so equivalent to what is being done. I updated `getenv` to `std::getenv` because the former is not thread safe even for concurrent read, and is easy to do.

If this change indeed fixes the crash, we can introduce a policy to remove all usages of `getenv` / `std::getenv` to avoid the issue in future, at least outside of `ray/util`. If it becomes necessary to have a low dependency way to read system environment variables like `TMPDIR`, we can revive the original approach in this PR then.

**Review by scv119 — APPROVED** [review]
Thanks for prioritizing this!

**Comment by scv119** [comment]
image

**Pull request merged**